### PR TITLE
[INF-986] Lease renew error callback

### PR DESCRIPTION
**Pull request by danpolyuha**

### CODE REVIEW

File: lib/settings_reader/vault_resolver/engines/abstract.rb [hunk 1]
```diff
@@ -50,18 +50,25 @@ def get_and_retry_connection(address)
...
         end
 
-        def renew_and_retry_auth(address)
-          renew_and_retry_connection(address)
-        rescue Vault::HTTPError => e # if not authenticated, let's reauthenticate and try once more
-          raise unless auth_error?(e)
+        def renew_and_retry_auth(entry)
+          renew_and_retry_connection(entry)
+        rescue Vault::HTTPError => e
+          return process_renew_auth_error(entry) if auth_error?(e)
```
Comment (volodymyr-mykhailyk) [inline lib/settings_reader/vault_resolver/engines/abstract.rb:56]: If I'm reading this code right way this is what going to happen:

1. Vault failover happens
2. Lease renew triggered
3. Auth error happens
4. Auth initiated
5. We retry renew
6. Lease not found, but an exception will not be caught as it happened outside of the rescue block

File: lib/settings_reader/vault_resolver/engines/abstract.rb [hunk 2]
```diff
@@ -50,18 +50,25 @@ def get_and_retry_connection(address)
...
 
-          config.vault_initializer.call
-          renew_and_retry_connection(address)
+          raise
+        end
+
+        private
+
+        def process_renew_lease_not_found_error(entry)
+          config.lease_not_found_handler.call(entry)
```
Comment (volodymyr-mykhailyk) [inline lib/settings_reader/vault_resolver/engines/abstract.rb:65]: Should we remove an entry from the cache as well here?

**Review by volodymyr-mykhailyk — CHANGES_REQUESTED** [review]
Thanks for the extension of functionality. However there are couple problems still exist:
- renewal problem will be reported only on the second renewal attempt after a connection was re-established. 
- we do not remove corrupted entry from cache - so new credentials will be retrieved only when existing expires completely
- old issue with entries renewal is still exist. Refresher class tries to renew lease on expired entries

**Comment by danpolyuha** [comment]
@volodymyr-mykhailyk All comments resolved, please take a look.

**Review by volodymyr-mykhailyk — APPROVED** [review]
🔥

**Pull request merged**